lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.js

```diff
@@ -198,32 +202,43 @@ function handleLogs() {
 
 function handleTags() {
   const provider = this.state.service.provider;
-  const tagsMerged = Object.assign({}, provider.stackTags, provider.tags);
-
-  const tagsToCreate = _.entriesIn(tagsMerged).map(pair => ({
-    key: String(pair[0]),
-    value: String(pair[1]),
-  }));
-  if (tagsToCreate) {
-    tagsToCreate.forEach((tag) => {
-      const operation = { op: 'replace', path: `/variables/${tag.key}`, value: tag.value };
-      this.apiGatewayStagePatchOperations.push(operation);
+  const tagsMerged = _.mapValues(Object.assign({}, provider.stackTags, provider.tags),
+      v => String(v));
+  const currentTags = this.apiGatewayStageState.tags || {};
+  const tagKeysToBeRemoved = Object.keys(currentTags)
+    .filter(currentKey => !_.isString(tagsMerged[currentKey]));
```

Comment:
Style suggestion: More to the point could be `tagsMerged.hasOwnProperty(currentKey)`, or if we created `tagsMerged` literal with `Object.create(null)` simply just `tagsMerged[currentKey]`.

(also`tagsMerged`could be setup as `Map`, still I get it turns cumbersome when we cannot rely on `Object.entries` and typical array utils, which are not out of a box available for maps)

Reply:
`~.hasOwnProperty` since it is forbidden by eslint rule `no-prototype-builtins`

We should not use `tagsMerged[currentKey]` to filter keys to be removed.
Because if tag value is a string `'0'` or `''` which is falsy, it wont be filtered.

`Object.assign` is used widely in codebase, since `Map` was not available on node.js prior to v4.
So I don't think there is a strong to choose `Map`. Will refactor in separate PR.

Reply:
> Unfortunately ~.hasOwnProperty cannot be used since it is forbidden by eslint rule no-prototype-builtins.

Yes, this rule is meant to address cases when we deal with externally passed objects, which are not guaranteed to have `.hasOwnProperty` (as e.g. ones created with `Object.create(null)`.

Here's it's not the case, but lint rule is not smart enough to detect that, so usually that's addressed with notation as:

```javascript
const hasOwnProperty = Object.prototype.hasOwnProperty;
...
hasOwnProperty.call(obj, propertyName)
```

Still here we could be cleaner if we create an object with `Object.create(null)`, as then we can simply rely on existence (or `currentKey in tagsMerged` if for some reason we care for `''`)

> We should not use tagsMerged[currentKey] to filter keys to be removed.
Because if a tag value is a string '0' or '' which is falsy, it wont be filtered.
So I used _.isString that does not care the content of string.

It's only `''` that's falsy and would be filtered. Still assuming we want to take in that as well, we can simply rely on `currentKey in tagsMerged`.

> So I don't think there is a strong preferrence to choose Map. Will refactor in separate PR.

Yes, no need, it was only a suggestion.